### PR TITLE
[SPARK-27204]SHS: Background caching for completed applications, when the disk store enabled

### PR DESCRIPTION
**Pull request by shahidki31**
## What changes were proposed in this pull request?
First time AppUI loading time will take time from the History server page, if the event log size is huge.
(Eg: 47 minutes for a 18GB eventlog file). Majority of time is consumed for replaying the event log.

In this pr, I'm proposing a background caching for disk store, where history provider will replay the event log once the application completed and creates a store cache. So, when the user try to access the application, they can directly get from the cache or, provider will wait till the time caching is completed.

Either case, the loading time will be lesser than or equal to the current one. Also, it is configurable and by default enabled. If disabled, behaviour will be same as before.

## How was this patch tested?
Manually tested and existing UT.
Event log size : 1.7 GB
**Before:**
 First time UI loading time = 391 sec 
**After:**
 First time UI loading time =  6 sec ( Best case: Already cached)
 First time UI loading time  =  344 sec (worst case: Immedietly after background caching started)

### CODE REVIEW

**Comment by shahidki31** [comment]
cc @vanzin , kindly give the feedback about the idea.
(code might need refactoring)

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 1]
```diff
@@ -405,6 +423,15 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
           executor.shutdownNow()
         }
       }
+
+      if (diskStoreCacheEnabled) {
+        Seq(replayPool.get).foreach { executor =>
```
Comment (felixcheung) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: `Seq(replayPool.get).foreach` -> `replayPool.foreach`

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 2]
```diff
@@ -122,6 +123,23 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
   // and applications between check task and clean task.
   private val pool = ThreadUtils.newDaemonSingleThreadScheduledExecutor("spark-history-task-%d")
 
+  private val diskStoreCacheEnabled = conf.get(LOCAL_STORE_DIR).nonEmpty &&
+    conf.get(DISK_BACKGROUND_CACHE_ENABLED)
+
+  // Create a threadpool for background disk caching, if enabled,
+  // "spark.history.fs.disk.backgroundCache.enabled"
+  private val replayPool = if (diskStoreCacheEnabled) {
```
Comment (felixcheung) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: this usage of something.get is kinda odd...

Comment (shahidki31) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Thanks @felixcheung. Yeah. I would like to know if the approach is okay, code wise still need some clean up.

**Pull request closed without merging**